In [1]:
import pandas as pd
import numpy as np
import ast

In [2]:
movies = pd.read_csv(r'/Users/rohitbandhakavi/Documents/ML Project/tmdb_5000_movies.csv')
credits = pd.read_csv(r'/Users/rohitbandhakavi/Documents/ML Project/tmdb_5000_credits.csv')

In [3]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [4]:
credits.head(1)['crew'].values

array(['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cam

In [5]:
movies = movies.merge(credits, on = 'title')

In [6]:
# checking the language distirbution of the movies
movies['original_language'].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
ta       2
cs       2
ro       2
id       2
ar       2
vi       1
sl       1
ps       1
no       1
ky       1
hu       1
pl       1
af       1
nb       1
tr       1
is       1
xx       1
te       1
el       1
Name: count, dtype: int64

Here the data is very skewed towards only english so the Model may be bad, so we are going to remove it.

In [7]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [8]:
# we are only keeping the following columns:
# generes
# id
# keywords
# title
# cast
# crew
# over view
movies = movies[['genres', 'id', 'overview', 'keywords', 'title', 'cast', 'crew']]

In [9]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   genres    4809 non-null   object
 1   id        4809 non-null   int64 
 2   overview  4806 non-null   object
 3   keywords  4809 non-null   object
 4   title     4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
dtypes: int64(1), object(6)
memory usage: 263.1+ KB


In [10]:
# adding the geners, keywords, cast, crew to the overview column

In [11]:
movies.isnull().sum()

genres      0
id          0
overview    3
keywords    0
title       0
cast        0
crew        0
dtype: int64

In [12]:
# removing the null rows
movies.dropna(inplace = True)

In [13]:
# no duplicates
movies.duplicated().sum()

0

In [14]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

the probelem here is that the genres is a string of the whole list as you can see below.
so we convert it into a list only ny using the 'ast' module 

In [15]:
type('genres')

str

In [16]:
# cow converting the genres column into a list of just the names and not the ID

def convert(obj):
    l = []
    for i in ast.literal_eval(obj):
        l.append(i['name'])
    return l

In [17]:
movies['genres'] = movies['genres'].apply(convert)

In [18]:
movies['keywords'] = movies['keywords'].apply(convert)

In [19]:
movies.head(1)

genres     id  \
0  [Action, Adventure, Fantasy, Science Fiction]  19995   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [20]:
#now we will try and take out only the first 3 cast members 


def convert2(obj):
    l = []
    count = 0
    for i in ast.literal_eval(obj):
        if count != 3:
            l.append(i['name'])
            count += 1
        else: 
            break
        
    return l

In [21]:
movies['cast'].apply(convert2)

0        [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1           [Johnny Depp, Orlando Bloom, Keira Knightley]
2            [Daniel Craig, Christoph Waltz, Léa Seydoux]
3            [Christian Bale, Michael Caine, Gary Oldman]
4          [Taylor Kitsch, Lynn Collins, Samantha Morton]
                              ...                        
4804    [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...
4805         [Edward Burns, Kerry Bishé, Marsha Dietlein]
4806           [Eric Mabius, Kristin Booth, Crystal Lowe]
4807            [Daniel Henney, Eliza Coupe, Bill Paxton]
4808    [Drew Barrymore, Brian Herzlinger, Corey Feldman]
Name: cast, Length: 4806, dtype: object

In [22]:
movies['cast'] = movies['cast'].apply(convert2)

In [23]:
movies.head(1)

genres     id  \
0  [Action, Adventure, Fantasy, Science Fiction]  19995   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [24]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

so from the many dictionaries in crew we will have to take only the dictionary where the job is director and then in that we will have to select the name only 

In [25]:
def fetch_director(obj):
    l = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            l.append(i['name'])
            break
    return l 

In [26]:
movies['crew'].apply(fetch_director)

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4806, dtype: object

In [27]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [28]:
movies.head(1)

genres     id  \
0  [Action, Adventure, Fantasy, Science Fiction]  19995   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [29]:
# converting the overview column into a string so that we can concatenate it with the other columns
movies['overview'].apply(lambda x:x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [30]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [31]:
movies.head()

genres      id  \
0  [Action, Adventure, Fantasy, Science Fiction]   19995   
1                   [Adventure, Fantasy, Action]     285   
2                     [Action, Adventure, Crime]  206647   
3               [Action, Crime, Drama, Thriller]   49026   
4           [Action, Adventure, Science Fiction]   49529   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

Since we are creating Tags for oyr model to recommend the movies to us we nee to remove the whitespaces bw the names or the genersetc.
like EX: in cast Tom Hanks will become TomHanks.
we do this because we dont want our model to confuse in our recommendation where if i like tom hanks movies what if the model recorganizes the Tom in tom cruise and shows me his movies. so we are doing this operation.
so using a lambda function to replace any white space with no space.

In [32]:
movies['genres'].apply(lambda x: [i.replace(' ', '') for i in x])  # you can see that Science Fiction became ScienceFiction

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [33]:
movies['genres'] = movies['genres'].apply(lambda x: [i.replace(' ', '') for i in x])
movies['keywords'] = movies['keywords'].apply(lambda x: [i.replace(' ', '') for i in x])
movies['cast'] = movies['cast'].apply(lambda x: [i.replace(' ', '') for i in x])
movies['crew'] = movies['crew'].apply(lambda x: [i.replace(' ', '') for i in x])

In [34]:
movies.head()

genres      id  \
0  [Action, Adventure, Fantasy, ScienceFiction]   19995   
1                  [Adventure, Fantasy, Action]     285   
2                    [Action, Adventure, Crime]  206647   
3              [Action, Crime, Drama, Thriller]   49026   
4           [Action, Adventure, ScienceFiction]   49529   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

creating a new tags column which will include overview, genres, keywords, cast, crew

In [35]:
movies['tags'] = movies['overview']+movies['genres']+movies['keywords']+movies['cast']+movies['crew']

In [36]:
movies.head()

genres      id  \
0  [Action, Adventure, Fantasy, ScienceFiction]   19995   
1                  [Adventure, Fantasy, Action]     285   
2                    [Action, Adventure, Crime]  206647   
3              [Action, Crime, Drama, Thriller]   49026   
4           [Action, Adventure, ScienceFiction]   49529   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [37]:
#creating a new DF

new_df = movies[[ 'id', 'title', 'tags']]

In [38]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [39]:
# converting tags to a string.
new_df['tags'].apply(lambda x: ' '.join(x))

0       In the 22nd century, a paraplegic Marine is di...
1       Captain Barbossa, long believed to be dead, ha...
2       A cryptic message from Bond’s past sends him o...
3       Following the death of District Attorney Harve...
4       John Carter is a war-weary, former military ca...
                              ...                        
4804    El Mariachi just wants to play his guitar and ...
4805    A newlywed couple's honeymoon is upended by th...
4806    "Signed, Sealed, Delivered" introduces a dedic...
4807    When ambitious New York attorney Sam is sent t...
4808    Ever since the second grade when he first saw ...
Name: tags, Length: 4806, dtype: object

In [40]:
new_df['tags'] = new_df['tags'].apply(lambda x: ' '.join(x))

/var/folders/6v/mlznxrkd0fq5q20j19qpkypm0000gn/T/ipykernel_3236/895399174.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: ' '.join(x))


In [41]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [42]:
new_df['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [43]:
# its recommended to convert everything to lowercase so we will do it

new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())

/var/folders/6v/mlznxrkd0fq5q20j19qpkypm0000gn/T/ipykernel_3236/3250951641.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())


If you see theat thre are multiple with past present and future sence like love, loved, loving.  so to club all those words into one we use stemming which will make all those words as one like love. For this we use nltk library which is nltk.

In [44]:
!pip install nltk

In [45]:
import nltk
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()   # we do this for ease of use.

In [46]:
def stem(text):
    y = []

    for i in text.split():    # doing this to convert the string to list so that each word can be stemmed.
        y.append(ps.stem(i))

    return " ".join(y)

In [47]:
# Example of stemming

ps.stem('loved')

'love'

In [48]:
new_df['tags'] = new_df['tags'].apply(stem)

/var/folders/6v/mlznxrkd0fq5q20j19qpkypm0000gn/T/ipykernel_3236/3213734980.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(stem)


In [49]:
new_df['tags'][0]

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

Now for the recommendation process we need to get the similarities bw tags so that id a person likes mission impossible then we can recommend him similar movies that is the most similar tags to that.
for that we are converting the string('tags') to vectors so that we can be able to plot them on the graph.
we are using text vectorization technique called bag of words.

In the bag of words model, we combine all the tags and lets say we have 5L words. then we take the n number od most used words(lets say n = 5000). now our thing is to creat a DF with all these most frequent vectors and the score of the appreance of these vectors in the tags of every single movie.
if we plot these scores in the graph then we can see that theres a movie A and then we can see the 5 nearest movies to that and recommend them to the user.
that means here every movie will become a vector in a n-dimensional space.
we are ot considering any stop words here cuz they dont make any sence to consider them.

In [50]:
# importing scikit learn module for the vectorization.
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 5000, stop_words = 'english')

In [51]:
vectors = cv.fit_transform(new_df['tags']).toarray() 

We use toarray cuz normally transform from sklearn returns a sparse matrix which only stores the position and the value of the non zero numbers in the matrix. So if we convert it to a numpy array(which is what.toarray does), then we store the whole matrix insted of the nonzero values.

In [52]:
np.set_printoptions(threshold=np.inf)  # this will make us see the full output.
cv.get_feature_names_out()

array(['000', '007', '10', '100', '11', '12', '13', '14', '15', '16',
       '17', '17th', '18', '18th', '18thcenturi', '19', '1910', '1920',
       '1930', '1940', '1944', '1950', '1950s', '1960', '1960s', '1970',
       '1970s', '1971', '1974', '1976', '1980', '1985', '1990', '1999',
       '19th', '19thcenturi', '20', '200', '2003', '2009', '20th', '21st',
       '23', '24', '25', '30', '300', '3d', '40', '50', '500', '60', '70',
       '80', 'aaron', 'aaroneckhart', 'abandon', 'abduct',
       'abigailbreslin', 'abil', 'abl', 'aboard', 'abov', 'abus',
       'academ', 'academi', 'accept', 'access', 'accid', 'accident',
       'acclaim', 'accompani', 'accomplish', 'account', 'accus', 'ace',
       'achiev', 'acquaint', 'act', 'action', 'actionhero', 'activ',
       'activist', 'activities', 'actor', 'actress', 'actual', 'ad',
       'adam', 'adamsandl', 'adamshankman', 'adapt', 'add', 'addict',
       'adjust', 'admir', 'admit', 'adolesc', 'adopt', 'ador',
       'adrienbrodi', 'adu

In [53]:
vectors.shape

(4806, 5000)

In [54]:
print(vectors)

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)



In this program, we use cosine distance to measure similarity between 5000-dimensional movie embeddings. Cosine distance focuses on the angle between vectors, which reflects how similar their directions are, regardless of their magnitude. This is important because two movies can have similar themes or genres (i.e., similar direction) even if their feature magnitudes differ. Euclidean distance, on the other hand, is sensitive to magnitude and performs poorly in high-dimensional spaces due to the curse of dimensionality. Therefore, cosine distance provides more meaningful and reliable results for content-based movie recommendations.

In [55]:
from sklearn.metrics.pairwise import cosine_similarity

In [56]:
similarity = cosine_similarity(vectors)

In [57]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , 0.07347184, 0.18929941,
       0.10838875, 0.04024218, 0.1467348 , 0.05923489, 0.09673017,
       0.10259784, 0.0946497 , 0.09037128, 0.04499213, 0.12824729,
       0.06282809, 0.07894737, 0.13977654, 0.09493291, 0.0830813 ,
       0.0580381 , 0.1096817 , 0.06622662, 0.08740748, 0.05333807,
       0.05101628, 0.15389675, 0.18693292, 0.11654331, 0.06503325,
       0.06684848, 0.15907119, 0.08520286, 0.09733285, 0.        ,
       0.09933993, 0.17316974, 0.07894737, 0.08111071, 0.08226127,
       0.07694838, 0.16563698, 0.        , 0.09086217, 0.0338255 ,
       0.08240856, 0.13910372, 0.19672237, 0.08447772, 0.05827165,
       0.1129565 , 0.08048436, 0.14843121, 0.0429735 , 0.03121953,
       0.07038153, 0.13834289, 0.02418254, 0.06748819, 0.07254763,
       0.01836796, 0.23179316, 0.08377078, 0.08175191, 0.02961744,
       0.01952916, 0.02249606, 0.16452255, 0.12049505, 0.02649065,
       0.07600419, 0.0733674 , 0.13710212, 0.03311331, 0.20246

here the similarity[0] means the similarity of the movie 1 with the rest of the 4806 movies. and since the movie 1 is exactly same to the first movie, so thats why the similarity score is 1.

In [58]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]  #fetching the index of the movie that is given in the function
    distance = similarity[movie_index]   # this fetched the similarity score of the movie with the index.
    movie_list = sorted(list(enumerate(distance)), reverse= True, key = lambda x:x[1])[1:6]

    for i in movie_list:
        print(new_df.iloc[i[0]].title)
  

This code finds and prints the top 5 movies most similar to a given movie based on a precomputed similarity matrix. First, it gets the index of the movie from the DataFrame new by matching the movie title

Then it retrieves the similarity scores between this movie and all others, pairs each score with its index using enumerate, and sorts them in descending order based on the similarity value:

Finally, it loops through the top 5 most similar movies (excluding the original movie, which is always most similar to itself) and prints their titles using their indices from the sorted list

key = lambda x: x[1] tells the sorted() function to sort based on the second element of each tuple. the enumerate op will be (movie indec, similarity score)
In this case, each item is a tuple like (movie_index, similarity_score), so x[1] is the similarity score.
This ensures the sorting is done by similarity values, not by the index.



In [59]:
recommend('Avatar') # these are the recommended movies.

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.


In [60]:
import pickle

In [61]:
pickle.dump(new_df, open('movie.pkl', 'wb'))

In [62]:
pickle.dump(similarity, open('similarity.pkl', 'wb'))